components/modules/net.c

```diff
@@ -1017,40 +1031,89 @@ static void lconnected_cb (lua_State *L, lnet_userdata *ud) {
   }
 }
 
-static void laccept_cb (lua_State *L, lnet_userdata *ud, struct tcp_pcb *newpcb) {
+static void laccept_cb (lua_State *L, lnet_userdata *ud) {
+  SYS_ARCH_DECL_PROTECT(lev);
   lua_rawgeti(L, LUA_REGISTRYINDEX, ud->server.cb_accept_ref);
 
   lnet_userdata *nud = net_create(L, TYPE_TCP_CLIENT);
   lua_pushvalue(L, -1);
   nud->self_ref = luaL_ref(L, LUA_REGISTRYINDEX);
-  nud->tcp_pcb = newpcb;
-  tcp_arg(nud->tcp_pcb, nud);
-  tcp_err(nud->tcp_pcb, net_err_cb);
-  tcp_recv(nud->tcp_pcb, net_tcp_recv_cb);
-  tcp_sent(nud->tcp_pcb, net_sent_cb);
-  nud->tcp_pcb->so_options |= SOF_KEEPALIVE;
-  nud->tcp_pcb->keep_idle = ud->server.timeout * 1000;
-  nud->tcp_pcb->keep_cnt = 1;
-  tcp_accepted(ud->tcp_pcb);
+
+  int recvevent = 0;
+  struct netconn *newconn;
+  err_t err = netconn_accept(ud->netconn, &newconn);
+  if (err == ERR_OK) {
+    nud->netconn = newconn;
+
+    SYS_ARCH_PROTECT(lev);
+    // take buffered receive events
+    recvevent = (int)(-1 - newconn->socket);
+    nud->netconn->socket = (int)nud;
+    SYS_ARCH_UNPROTECT(lev);
+
+    netconn_set_nonblocking(nud->netconn, 1);
+    netconn_set_noautorecved(nud->netconn, 1);
+    nud->netconn->pcb.tcp->so_options |= SOF_KEEPALIVE;
+    nud->netconn->pcb.tcp->keep_idle = ud->server.timeout * 1000;
+    nud->netconn->pcb.tcp->keep_cnt = 1;
+  } else
+    luaL_error(L, "cannot accept new server socket connection");
 
   lua_call(L, 1, 0);
+
+  while (recvevent-- > 0) {
+    // kick receive callback in case of pending events
+    post_net_recv(nud);
+  }
 }
 
-static void lrecv_cb (lua_State *L, lnet_userdata *ud, const lnet_recvdata *rd) {
-  if (ud->client.cb_receive_ref != LUA_NOREF) {
+static void lrecv_cb (lua_State *L, lnet_userdata *ud) {
+  if (!ud->netconn) return;
+
+  struct netbuf *p;
+  char *payload;
+  uint16_t len;
+
+  err_t err = netconn_recv(ud->netconn, &p);
+  if (err != ERR_OK) {
+    lwip_lua_checkerr(L, err);
+    return;
+  }
+  if (p) {
+    netbuf_data(p, (void **)&payload, &len);
+  } else {
+    len = 0;
+  }
+
+  if (len > 0 && ud->client.cb_receive_ref != LUA_NOREF){
     lua_rawgeti(L, LUA_REGISTRYINDEX, ud->client.cb_receive_ref);
     int num_args = 2;
     lua_rawgeti(L, LUA_REGISTRYINDEX, ud->self_ref);
-    lua_pushlstring(L, rd->payload, rd->payload_len);
+    lua_pushlstring(L, payload, len);
     if (ud->type == TYPE_UDP_SOCKET) {
       num_args += 2;
       char iptmp[IP_STR_SZ];
-      ipstr (iptmp, &rd->src_ip);
-      lua_pushinteger(L, rd->src_port);
+      ip_addr_t *addr = netbuf_fromaddr(p);
+      uint16_t port = netbuf_fromport(p);
+      ipstr (iptmp, addr);
+      lua_pushinteger(L, port);
       lua_pushstring(L, iptmp);
     }
     lua_call(L, num_args, 0);
   }
+
+  if (p) {
+    netbuf_delete(p);
+
```

Comment:
The latter. `netbuf_data()` just provides the data pointer and length of the embedded pbuf, while `netbuf_delete()` frees both the netbuf and the pbuf.